Add support for builtin_verbose_trap

ahatanak opened the pull request:
The builtin causes the program to stop its execution abnormally and shows a human-readable description of the reason for the termination when a debugger is attached or in a symbolicated crash log.

The motivation for the builtin is explained in the following RFC:

https://discourse.llvm.org/t/rfc-adding-builtin-verbose-trap-string-literal/75845

clang's CodeGen lowers the builtin to `llvm.trap` and emits debugging information that represents an artificial
inline frame whose name encodes the category and reason strings passed to the builtin.

ahatanak committed:
Add support for builtin_verbose_trap

The builtin causes the program to stop its execution abnormally and shows a
human-readable description of the reason for the termination when a debugger is
attached or in a symbolicated crash log.

The motivation for the builtin is explained in the following RFC:

https://discourse.llvm.org/t/rfc-adding-builtin-verbose-trap-string-literal/75845

ahatanak committed:
Fix diagnostic message

adrian-prantl commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -1628,6 +1628,27 @@ llvm::DIType *CGDebugInfo::createFieldType(
 }
 
+llvm::DISubprogram *
+CGDebugInfo::getFakeFuncSubprogram(const std::string &FakeFuncName) {
```
Maybe call this `createInlinedTrapSubprogram`?

adrian-prantl commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -3416,6 +3437,27 @@ llvm::DIMacroFile *CGDebugInfo::CreateTempMacroFile(llvm::DIMacroFile *Parent,
 
+llvm::DILocation *CGDebugInfo::CreateTrapFailureMessageFor(
+    llvm::DebugLoc TrapLocation, StringRef Prefix, StringRef FailureMsg) {
+  // Create debug info that describes a fake function whose name is the failure
```
```suggestion
  // Create debug info that describes an inlined function whose name is the error
```

adrian-prantl reviewed (commented):
A cosmetic suggestion I have would be to replace all instances of `Fake` with `inlinedTrap` to make the purpose clearer.

dwblaikie commented on clang/test/SemaCXX/verbose-trap.cpp:
```diff
@@ -0,0 +1,28 @@
+// RUN: %clang_cc1 -std=c++11 -fsyntax-only -fcxx-exceptions -verify %s
+
+#if !__has_builtin(__builtin_verbose_trap)
+#error
+#endif
```
Is this necessary?

ahatanak replied:
This just makes sure the builtin is supported unconditionally and doesn't depend on some features or anything (cc @ldionne ).

dwblaikie reviewed (commented):
neat idea (I imagine some debuggers will trip over the space in the inlined name, but seems pretty good)

I'd guess sometimes we'll lose the inlined debug info if two of these traps are merged, fwiw - but it's probably relatively robust, apart from that.

adrian-prantl commented:
For context, we've been using this same trick in the Swift standard library for a long time now, to get a good trade-off between space-efficient code generation and usability for inlineable functions such as the arithmetic `+` operator, which traps on overflow.

adrian-prantl commented:
> neat idea (I imagine some debuggers will trip over the space in the inlined name, but seems pretty good)

For spaces specifically there's a multi-decade precedent: Objective-C methods have names like `DW_AT_name ("-[NSObject init:]")`

felipepiovezan commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -3416,6 +3437,27 @@ llvm::DIMacroFile *CGDebugInfo::CreateTempMacroFile(llvm::DIMacroFile *Parent,
+  }
+
+  assert(FuncName.size() > 0);
+  assert(FuncName.find(' ') != std::string::npos &&
```
the pre condition (as stated in the function documentation) is that `Prefix` should have a space.
But we add a space whenever `FailureMsg` is not empty, so this assert is not catching a pre condition violation in all cases

ahatanak replied:
I think it means `Prefix` is required to have a space only when `FailureMsg` is empty. If `FailureMsg` is empty, `FuncName` is the same as `Prefix`, so the checks seem correct to me. @delcypher is that correct?

felipepiovezan replied:
We can better express this intent by moving the assert _before_ the `if` and by rewriting it as the faster `assert(!FailureMessage.empty() || FuncName.find(' ') != npos)`

ahatanak replied:
I think that should be fine. We should also make sure someone doesn't accidentally remove the space in `FuncName` inside the `if`, but I guess that's okay since there's a comment there.

adrian-prantl replied:
I think the right answer here is:
We guarantee not to collide with other symbols by use the reserved `__` prefix. Any consumer that wants to identify such a symbol checks for the `__llvm_builtin_trap` prefix to unambiguously identify these functions.

adrian-prantl replied:
The space is not needed for correctness.

pogo59 commented on clang/docs/LanguageExtensions.rst:
```diff
@@ -3379,6 +3379,54 @@ Query for this feature with ``__has_builtin(__builtin_debugtrap)``.
+**Description**
+
+``__builtin_verbose_trap`` is lowered to the ` ``llvm.trap`` <https://llvm.org/docs/LangRef.html#llvm-trap-intrinsic>`_ builtin.
+Additionally, clang emits debug metadata that represents an artificial inline
```
"metadata" is a clang internal term, of no real meaning to end users, and this page is intended for end users. I think "debugging information" would be better.

pogo59 commented on clang/docs/LanguageExtensions.rst:
```diff
@@ -3379,6 +3379,54 @@ Query for this feature with ``__has_builtin(__builtin_debugtrap)``.
+        __builtin_verbose_trap("Argument_must_not_be_null");
+    }
+
+The debug metadata would look as if it were produced for the following code:
```
"debugging information"

pogo59 commented on clang/docs/LanguageExtensions.rst:
```diff
@@ -3379,6 +3379,54 @@ Query for this feature with ``__has_builtin(__builtin_debugtrap)``.
+    }
+
+However, the LLVM IR would not actually contain a call to the artificial
+function — it only exists in the debug metadata.
```
However, the generated code would not ... in the debugging information.

delcypher replied:
@ahatanak Any follow up?

pogo59 reviewed (commented):
The documentation should say that this works only when debug information is enabled.

ldionne commented on clang/test/SemaCXX/verbose-trap.cpp:
```diff
@@ -0,0 +1,28 @@
+
+template <const char * const str>
+void f(const char * arg) {
+  __builtin_verbose_trap("Argument_must_not_be_null");
```
```suggestion
  __builtin_verbose_trap("Arbitrary string literals can be used!");
  __builtin_verbose_trap("Argument_must_not_be_null");
```

I think this would require encoding the message, but it's really important to be able to pass arbitrary string literals here.

var-const replied:
+1. Can we also check:
1. A non-ASCII Unicode character?
2. A long string (is there any limitation on the string length?)?

ahatanak replied:
1. Yes, it can handle something like `__builtin_verbose_trap(u8"¡¢£¤¥¦§¨©ª");`.
2. How long of a string do you need to test? I don't think there's a limitation that's specific to this builtin.

ldionne replied:
Oh, okay, then I don't understand how this works under the hood, but I'm happy :)

ahatanak replied:
I noticed that the code doesn't compile in `c++20` because the type of UTF8 string literals is `char8_t[N]` instead of `char[N]`. This is the error message:

`cannot initialize a parameter of type 'const char *' with an lvalue of type 'const char8_t[21]'`

It compiles if I cast it to the correct type (`__builtin_verbose_trap((const char *)u8"¡¢£¤¥¦§¨©ª");`).

How important is it to be able to pass UTF8 string literals to the builtin in c++20 without using any casts?

ahatanak replied:
It looks like we'd need another version of the builtin (`__builtin_verbose_trap_8t`?) that accepts `const char8_t *` if we wanted to pass UTF8 string literals in c++20.

pogo59 replied:
You can't overload the builtin? IIRC there are others that do this, although I can't name one offhand.

ldionne replied:
FWIW I think not being able to pass actual UTF8 is reasonable. When I said "arbitrary string literal", I meant anything that you can put in a normal `char` string literal, like spaces and special characters.

ahatanak replied:
@pogo59,  it looks like it's possible to overload builtins using templates in `clang/Basic/Builtins.td` as you said (for example, see `SyncFetchAndAddN`). However,

1. `char8_t` for builtin types isn't currently supported. We'll need to fix a few places in clang including tablegen to support that type.
2. It seems that the overloaded builtin will need a suffix (e.g., `_u8`) to distinguish it from the original one.

I think we can try to add support for that in a follow up patch unless it's something we need now.

philnik777 replied:
You can overload builtins by adding custom type checking to them. The "templates" in `Builtins.td` is more of a way to generate builtins with the same pattern for different types. e.g. there are `__builtin_cos`, `__builtin_cosf` and `__builtin_cosl`, which are essentially identical, except that they take `double`, `float` and `long double` respectively. To avoid typing everything out (and making errors that way), you can make the builtin a template.

var-const replied:
IMO it's completely reasonable to leave supporting Unicode as a potential follow-up, but I'm happy to know what currently is and isn't supported.

For the string length -- maybe let's try 128 or 256 characters? There isn't any specific size that we need to test -- just want some validation that passing a string longer than a short sentence won't cause any weird effects.

ahatanak replied:
@philnik777 you are right. I'll enable unicode in follow-up patches using custom type checking.

@var-const 128 and 256 character strings should work without any problems. I don't think you'll see any problems with long strings unless you hit implementation limits of clang or lldb. I tested a 512 char string literal locally and both clang and lldb work fine.

delcypher replied:
Imposing this length limit seems pretty arbitrary.  What's the actually restriction on the debug info/LLDB side? If there isn't one I don't think we should be adding an artificial limit without a good reason for it.

delcypher replied:
As long as the compiler isn't imposing an artificial limit then that seems fine to me.

dwblaikie replied:
FWIW, I've seen multiple MB symbol names :) so I probably wouldn't be too worried that any user would write a description that's longer than the symbols that complex templates can create...

ahatanak committed:
Address review comments

var-const commented on clang/lib/CodeGen/CGDebugInfo.h:
```diff
@@ -346,6 +346,14 @@ class CGDebugInfo {
 
+  // A cache that maps artificial inlined function names used for
+  // __builtin_verbose_trap to subprograms.
+  std::map<std::string, llvm::DISubprogram *> InlinedTrapFuncMap;
```
Nit: I think this header might need to include `<map>` and `<string>`, otherwise it appears to rely on those headers being included transitively from somewhere. (Unless of course there's some Clang convention against doing that)

dwblaikie replied:
& consider using `llvm::StringMap`, perhaps?

var-const reviewed (commented):
From the libc++ perspective, the only substantial comment I have is that ideally we need to make sure it's possible to pass long strings with arbitrary characters as the string parameter (or, if it's infeasible to implement, we need to document the limitations).

ldionne commented on clang/test/SemaCXX/verbose-trap.cpp:
```diff
@@ -0,0 +1,29 @@
+  __builtin_verbose_trap("hello" "world");
+  __builtin_verbose_trap(constMsg1);
+  __builtin_verbose_trap(constMsg2);
+  __builtin_verbose_trap(""); // expected-error {{argument to __builtin_verbose_trap must be a pointer to a non-empty constant string}}
```
Would it be possible to make that (an empty string literal) work? It seems like a bit of an arbitrary limitation.

ahatanak replied:
Yes, it's possible.

var-const reviewed (approved):
LGTM from the libc++ perspective, thanks a lot for working on this!

ahatanak committed:
Address review comments

1. Test long string argument.
2. Test c++20 unicode.
3. Allow empty string arguments.
4. #inlcude <map> and <string>.

ahatanak committed:
Add triple to test

dwblaikie commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -3424,6 +3445,26 @@ llvm::DIMacroFile *CGDebugInfo::CreateTempMacroFile(llvm::DIMacroFile *Parent,
+
+  // Create debug info that describes an inlined function whose name is the
+  // failure message.
+  std::string FuncName(Prefix);
```
`llvm::SmallString`?

dwblaikie commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -3424,6 +3445,26 @@ llvm::DIMacroFile *CGDebugInfo::CreateTempMacroFile(llvm::DIMacroFile *Parent,
   return DBuilder.createTempMacroFile(Parent, Line, FName);
 }
 
+llvm::DILocation *CGDebugInfo::CreateTrapFailureMessageFor(
```
There's currently only one caller with a fixed parameter for `Prefix` - perhaps that could be hardcoded in the implementation instead?

delcypher replied:
@dwblaikie We currently use this function for `-fbounds-safety` internally. We are currently in the process of upstreaming this which is why the function should **not** hardcode the `Prefix`.

ahatanak replied:
@dwblaikie what do you think? `-fbounds-safety` is currently being upstreamed and there is going to be another caller of the function in the not-too-distant future.

ahatanak replied:
@dwblaikie any comments?

dwblaikie replied:
I think I agree with https://github.com/llvm/llvm-project/pull/79230/files#r1529429173 then - the prefix should be a parameter and I don't think it needs to be a #defined constant (CLANG_VERBOSE_TRAP_PREFIX) - and can instead be written as a literal in the caller in CGBuiltin.cpp?

ahatanaka replied:
OK, I'll make it a parameter then.

I added macro `CLANG_VERBOSE_TRAP_PREFIX` to `ModuleBuilder.h` so that lldb wouldn't have to define it too. See https://github.com/llvm/llvm-project/pull/80368#discussion_r1481781470.

ahatanak replied:
@dwblaikie 's comment:
https://github.com/llvm/llvm-project/pull/79230#discussion_r1575391593

So the name of the subprogram would be something like `__builtin_trap: identifier123 : Argument_must_not_be_null`?

@Michael137 @delcypher is there a reason we cannot use the same prefix (`__builtin_trap`, for example)? Does lldb need clang to use different prefixes for `-fbounds-safety` and `builtin_verbose_trap`?

ahatanaka replied:
I've updated the patch to pass a category string to the builtin. The artificial function name has the format `<prefix>:<category>:<message>` where `<prefix>` is always `__builtin_verbose_trap`.

We should probably use something other than `:` for the separator as users might want to use it in the category or message string (e.g., `boost::some_library_name`).

ahatanak replied:
@dwblaikie any other comments?

dwblaikie replied:
Eh, I'm not too fussed about the separator for the prefix (we control that) and category (will the category be user-visible? Like a warning group? Or is that only still an implementation detail/contract between DWARF producer and DWARF consumer? I thought it was more the latter/implementation detail? In which case we can say what characters are valid there, and I'd keep it pretty simple, like the warning group names - all lower, dash or underscore separated seems fine)

ahatanak committed:
Address review comments

ahatanak committed:
Pass a better file scope when creating the subprogram for the artificial
inlined function

`TheCU->getFile()` returns `<stdin>` when `-main-file-name` is empty.

adrian-prantl commented on clang/lib/CodeGen/CGDebugInfo.h:
```diff
@@ -602,6 +613,19 @@ class CGDebugInfo {
+  //
+  // Currently `<Prefix>` is always "__llvm_verbose_trap".
+  //
+  // This is used to store failure reasons for traps.
```
///

delcypher replied:
@ahatanak Any follow up?

adrian-prantl commented on clang/lib/CodeGen/CGDebugInfo.h:
```diff
@@ -346,6 +348,15 @@ class CGDebugInfo {
       llvm::ArrayRef<llvm::Metadata *> PreviousFieldsDI, const RecordDecl *RD);
 
+  // A cache that maps artificial inlined function names used for
+  // __builtin_verbose_trap to subprograms.
```
///

delcypher replied:
@ahatanak Any follow up?

delcypher commented on clang/test/SemaCXX/verbose-trap.cpp:
```diff
@@ -0,0 +1,35 @@
+  __builtin_verbose_trap("");
+  __builtin_verbose_trap(); // expected-error {{too few arguments}}
+  __builtin_verbose_trap(0); // expected-error {{argument to __builtin_verbose_trap must be a pointer to a constant string}}
+  __builtin_verbose_trap(1); // expected-error {{cannot initialize a parameter of type 'const char *' with}}
```
Did this error message get truncated? Is this deliberate?

delcypher commented on clang/test/CodeGenCXX/debug-info-verbose-trap.cpp:
```diff
@@ -0,0 +1,49 @@
+// CHECK: ![[SUBPROG32:.*]] = distinct !DISubprogram(name: "f2<&constMsg[0]>", linkageName: "_Z2f2IXadsoKcL_ZL8constMsgEEEEvv",
+// CHECK: ![[LOC36]] = !DILocation(line: 0, scope: ![[SUBPROG26]], inlinedAt: ![[LOC37:.*]])
+// CHECK: ![[LOC37]] = !DILocation(line: 44, column: 3, scope: ![[SUBPROG32]])
+
```
Should this test check that the `llvm.trap()` function has the `cold` attribute (hint to the optimizer that its unlikely for the function to be called) on it?

delcypher reviewed (changes requested):
Looks pretty good. I have a few minor comments. Some things I suggest we may want to do in a follow up patch.

delcypher commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -3424,6 +3443,26 @@ llvm::DIMacroFile *CGDebugInfo::CreateTempMacroFile(llvm::DIMacroFile *Parent,
+  SmallString<64> FuncName(Prefix);
+  if (!FailureMsg.empty()) {
+    // A space in the function name identifies this as not being a real function
+    // because it's not a valid symbol name.
```
@adrian-prantl Sorry. This comment is based on an off-hand comment you made a while back in the context of `-fbounds-safety` (which only supports C) where I think you said having a space in the function name would distinguish the artificial function from any real function. While that's true for C I guess it's not true for other languages.

delcypher commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -3424,6 +3443,26 @@ llvm::DIMacroFile *CGDebugInfo::CreateTempMacroFile(llvm::DIMacroFile *Parent,
+                                         StringRef FailureMsg) {
+  // Create a debug location from `TrapLocation` that adds an artificial inline
+  // frame.
+  const char *Prefix = "__llvm_verbose_trap";
```
Could "__llvm_verbose_trap" live in a shared header file? That way LLDB doesn't need to hardcode the string in a frame recognizer implementation. It can just use the string in the header file and that way things stay in sync automatically.

ahatanak replied:
Yes, I think it can be in `clang/include/clang//CodeGen/ModuleBuilder.h` or another header in the directory.

delcypher commented on clang/lib/CodeGen/CGDebugInfo.h:
```diff
@@ -346,6 +348,15 @@ class CGDebugInfo {
 
+  // A cache that maps artificial inlined function names used for
+  // __builtin_verbose_trap to subprograms.
+  llvm::StringMap<llvm::DISubprogram *> InlinedTrapFuncMap;
```
Just to check. What's the life time of CGDebugInfo? Is it a single instance per complication unit? If it's longer than that (i.e. the same instance gets used for multiple compilation units) then we might cache the sub programs for too long.

ahatanak replied:
It doesn't live longer than a compilation unit. `CodeGenModule::DebugInfo` is created in the constructor of `CodeGenModule`.

ahatanak commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -3424,6 +3443,26 @@ llvm::DIMacroFile *CGDebugInfo::CreateTempMacroFile(llvm::DIMacroFile *Parent,
+                                         StringRef FailureMsg) {
+  // Create a debug location from `TrapLocation` that adds an artificial inline
+  // frame.
+  const char *Prefix = "__llvm_verbose_trap";
```
See https://github.com/llvm/llvm-project/pull/79230#discussion_r1471967969

Should we make it a parameter when we add support for `-fbounds-safety`?

ahatanak commented on clang/test/CodeGenCXX/debug-info-verbose-trap.cpp:
```diff
@@ -0,0 +1,49 @@
+// RUN: %clang_cc1 -triple arm64-apple-ios -std=c++20 -emit-llvm -debug-info-kind=limited %s -o - | FileCheck %s
```
Do we want to add a clang test? I thought we generally don't enable optimization in clang CodeGen tests because it makes the tests fragile. Has the guideline changed?

delcypher replied:
Really? We don't test the generated IR in an optimized build? That seems like a bad idea given that code built for production use typically is optimized.

delcypher replied:
Really? We don't test the generated IR in an optimized build? That seems like a bad idea given that code built for production use typically is optimized.

delcypher replied:
@ahatanak Any follow up?

pogo59 replied:
If you want to test optimization behavior, that would be an LLVM test starting from IR. Clang itself does not optimize and so should not test optimization behavior.

ahatanak replied:
https://llvm.org/docs/TestingGuide.html#best-practices-for-regression-tests recommends we run a minimal set of passes. As @pogo59 said, I think we should add a separate LLVM test if we want to test the merging behavior.

delcypher replied:
@pogo59 @ahatanak Thanks for the explanation.

A slight issue with that is some cases Clang will generate different (unoptimized IR) depending on the optimization level (see `CodeGenFunction::EmitTrapCheck`) so in those cases you have to test Clang's optimizing behavior. You're not using that here so that's fine

ahatanak replied:
In that case, we can use `-disable-llvm-passes` or `-disable-llvm-optzns` to avoid running the llvm optimization passes.

ahatanak commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -3452,6 +3452,18 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
+          getDebugInfo()->CreateTrapFailureMessageFor(TrapLocation, Str);
+    }
+    ApplyDebugLocation ApplyTrapDI(*this, TrapLocation);
+    EmitTrapCall(Intrinsic::trap);
```
I can tackle it in a separate patch.

For mode 2 and 3 calls, we can annotate the calls with something that indicates whether they can be merged only with other calls if the reason is the same or they cannot be merged at all. For calls in mode 3, we can just annotate the calls with attribute `nomerge`. For mode 2, I think we need to annotate the calls with an attribute or something that distinguishes the calls from the calls that are mergeable (mode 1) and the ones that aren't (mode 3).

delcypher replied:
Please leave a comment in the code indicating that currently no attempt is made to prevent traps being merged.

ahatanak commented on clang/test/CodeGenCXX/debug-info-verbose-trap.cpp:
```diff
@@ -0,0 +1,49 @@
+// CHECK: ![[FILESCOPE:.*]] = !DIFile(filename: "{{.*}}debug-info-verbose-trap.cpp"
+// CHECK: ![[SUBPROG14:.*]] = distinct !DISubprogram(name: "f0", linkageName: "_Z2f0v",
+// CHECK: ![[LOC17]] = !DILocation(line: 0, scope: ![[SUBPROG18:.*]], inlinedAt: ![[LOC20:.*]])
+// CHECK: ![[SUBPROG18]] = distinct !DISubprogram(name: "__llvm_verbose_trap: Argument_must_not_be_null", scope: ![[FILESCOPE]], file: ![[FILESCOPE]], type: !{{.*}}, flags: DIFlagArtificial, spFlags: DISPFlagDefinition, unit: !{{.*}})
```
Yes, the test checks that the subprograms with the same message aren't duplicated. `LOC17` and `LOC23` have the same scope `SUBPROG18`. Also, `LOC25` and `LOC36` have the same scope because both calls have the message `hello`.

ahatanak committed:
Address review comments

ahatanak committed:
Define CLANG_VERBOSE_TRAP_PREFIX in public header

delcypher commented on clang/lib/CodeGen/CGDebugInfo.cpp:
```diff
@@ -3424,6 +3447,24 @@ llvm::DIMacroFile *CGDebugInfo::CreateTempMacroFile(llvm::DIMacroFile *Parent,
+                                         StringRef FailureMsg) {
+  // Create a debug location from `TrapLocation` that adds an artificial inline
+  // frame.
+  const char *Prefix = CLANG_VERBOSE_TRAP_PREFIX;
```
The `Prefix` should not be hardcoded. It should be a parameter to `CreateTrapFailureMessageFor` so it can be re-used by other things like `-fbounds-safety`

delcypher reviewed (changes requested):
It looks like there are still some unresolved discussions in this PR.

delcypher commented on clang/docs/LanguageExtensions.rst:
```diff
@@ -3379,6 +3379,57 @@ Query for this feature with ``__has_builtin(__builtin_debugtrap)``.
+``__builtin_verbose_trap`` is lowered to the ` ``llvm.trap`` <https://llvm.org/docs/LangRef.html#llvm-trap-intrinsic>`_ builtin.
+Additionally, clang emits debugging information that represents an artificial
+inline frame whose name encodes the string passed to the builtin, prefixed by a
+"magic" prefix.
```
@ahatanak Any follow up?

delcypher commented on clang/docs/LanguageExtensions.rst:
```diff
@@ -3379,6 +3379,60 @@ Query for this feature with ``__has_builtin(__builtin_debugtrap)``.
+        "__llvm_verbose_trap: Argument must not be null!"();
+    }
+
+However, the LLVM IR would not actually contain a call to the artificial
```
```suggestion
However, the generated code would not actually contain a call to the artificial
```

delcypher commented on clang/lib/CodeGen/CGDebugInfo.h:
```diff
@@ -346,6 +348,15 @@ class CGDebugInfo {
+  llvm::StringMap<llvm::DISubprogram *> InlinedTrapFuncMap;
+
+  // A function that returns the subprogram corresponding to the artificial
+  // inlined function for traps.
```
@ahatanak Any follow up?

ahatanak committed:
Address review comments

ahatanak committed:
Use three forward slashes

delcypher commented on clang/docs/LanguageExtensions.rst:
```diff
@@ -3379,6 +3379,60 @@ Query for this feature with ``__has_builtin(__builtin_debugtrap)``.
 
 Query for this feature with ``__has_builtin(__builtin_trap)``.
 
+``__builtin_verbose_trap``
```
@ahatanak Is this in the wrong place in the document? `__builtin_verbose_trap` is lexicographically after `__builtin_nondeterministic_value`

ahatanak replied:
It doesn't look like the existing builtins are in lexicographical order. `__builtin_debugtrap` is followed by `__builtin_trap`, which is followed by `__builtin_nondeterministic_value`.

dwblaikie replied:
Ah, OK - that seems like more reason the prefix should be singular. Like we shouldn't add a new/distinct prefix for array bounds? Or other things we want to error out on - because it'd be an ongoing maintenance problem updating consumers like lldb each time we add one? 

So maybe we should have a generic prefix we'll always use (`__builtin_trap`, say) and then maybe some identifier after that in a known format (eg: space, identifier, space) that can be used? then an error message string?

ahatanak committed:
Fix test

ahatanak committed:
Pass category string to the builtin

ahatanak committed:
Use $ as the separator

ahatanak commented:
I'll merge this in a day or two if there are no objections.

delcypher commented on clang/include/clang/CodeGen/ModuleBuilder.h:
```diff
@@ -27,6 +27,9 @@ namespace llvm {
   }
 }
 
+// Prefix for __builtin_verbose_trap.
```
@ahatanak Does this comment need updating? The prefix isn’t just for __builtin_verbose_trap anymore.

delcypher commented on clang/include/clang/CodeGen/ModuleBuilder.h:
```diff
@@ -27,6 +27,9 @@ namespace llvm {
 }
 
+// Prefix for __builtin_verbose_trap.
+#define CLANG_VERBOSE_TRAP_PREFIX "__llvm_verbose_trap"
```
@ahatanak Nit: Why does the macro name mention `CLANG` but the string mentions `llvm` instead of Clang? Seems like this prefix is specific to Clang and not LLVM.

Also here’s suggestion for a shorter prefix `__llvm_trap_msg`. It’s not a big deal so feel to ignore this advice.

adrian-prantl replied:
Does this have to be a macro or could it be a C++ constant?

ahatanak replied:
If we used a C++ constant, wouldn't it be harder to make a compile time constant string (e.g., `^__llvm_verbose_trap: `)?

https://github.com/llvm/llvm-project/pull/80368/files#diff-e87fc223a3fc5310a2dc415999059d623bb89cd10f03324ac4c4496cd7f5419f

ahatanak replied:
@delcypher I don't think it has to mention `llvm`. Can we make it even shorter, e.g., `__trap_msg`?

dwblaikie replied:
I think we usually use some llvm or clang prefix in these sort of things, to reduce the chance of conflict with some other implementation's reserved names? Could be worth a quick grepping for other similar symbol names in clang/llvm to see what we do/if there's some pattern to follow?

dwblaikie replied:
& not sure it being a compile time constant regex is super important - the regex compilation probably dwarfs the string concatenation required to add the `^` and `:`?

ahatanak replied:
I can make it a C++ constant if that's preferable.

ahatanak replied:
I don't have any preference, but according to people working on lldb, macro would make it marginally easier to construct the regex in the LLDB frame recognizer.

Michael137 replied:
Not a big deal for LLDB either way, whatever people prefer here, we can work around it.

ahatanak committed:
Use a shorter prefix

ahatanak commented:
We could use the function name without the prefix as the key when searching for the subprogram in `CGDebugInfo::createInlinedTrapSubprogram`.

adrian-prantl commented on clang/lib/CodeGen/CGDebugInfo.h:
```diff
@@ -29,7 +29,9 @@
 #include "llvm/IR/DebugInfo.h"
 #include "llvm/IR/ValueHandle.h"
 #include "llvm/Support/Allocator.h"
+#include <map>
```
This seems to not belong to this commit?

adrian-prantl commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -172,6 +172,27 @@ static bool checkArgCount(Sema &S, CallExpr *Call, unsigned DesiredArgCount) {
+      continue;
+
+    // FIXME: Add more checks and reject strings that can't be handled by
+    // debuggers.
```
I don't think we need this FIXME. A debug info producer can't know what can and cannot be handled by debuggers, and DWARF consumers need to support any string in the specified encoding (usually UTF-8) anyway.

delcypher commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -172,6 +172,27 @@ static bool checkArgCount(Sema &S, CallExpr *Call, unsigned DesiredArgCount) {
          << /*is non object*/ 0 << Call->getArg(1)->getSourceRange();
 }
 
+static bool checkBuiltinVerboseTrap(CallExpr *Call, Sema &S) {
```
@ahatanak Now that we've changed the interface I think this Sema check should be slightly different.

* We should be checking that the number of arguments is **exactly** 2. This looks like it allows any number.
* The `Category` argument should not be allowed to contain the separator character (`$`).

ahatanak replied:
clang already checks that the number of arguments is exactly 2 (see `test/SemaCXX/verbose-trap.cpp`).

I agree that we should check that `$` isn't in the strings.

dwblaikie commented:
Hmm, do other builtins actually end up as symbol names? I think most of them lower to an instruction or two - I guess this one doesn't lower to an actual symbol, only a DWARF symbol - but maybe that's still different enough it should use an llvm or clang in the name... (sorry, I think I saw this was discussed somewhere on the review, but I can't find where now... :/)

I guess it came up here: https://github.com/llvm/llvm-project/pull/79230#discussion_r1591335258 and at the time it still had `llvm` in it, wasn't clear conversation that showed how the alternative conclusion was reached.

var-const commented:
@dwblaikie @ahatanak Is there anything we can do to move this patch forward? It seems like it's almost ready to go with perhaps one remaining thread of conversation to address. We badly need this functionality in libc++, the lack of error messages when a hardening check is triggered results in very poor user experience. It also makes it harder to talk about hardening since we essentially need to describe both the current state and the state after this patch is merged.

dwblaikie commented:
I think my last comment/question is still open? How/why did the symbol name end up dropping any llvm/clang component to avoid collisions with other names?

ahatanak commented:
> I think my last comment/question is still open? How/why did the symbol name end up dropping any llvm/clang component to avoid collisions with other names?

I dropped llvm/clang because I didn't think it would cause any collision in practice, but I don't think we have to try too hard to shorten the prefix.

ahatanak committed:
Use  as the prefix

ahatanak committed:
Reject arguments using '$'

dwblaikie commented on clang/include/clang/CodeGen/ModuleBuilder.h:
```diff
@@ -27,6 +27,9 @@ namespace llvm {
 }
 
+// Prefix of the name of the artificial inline frame.
+#define CLANG_TRAP_PREFIX "__clang_trap_msg"
```
inline StringRef? usual reasons to avoid macros, etc

dwblaikie reviewed (approved):
Looks OK - one minor nit, I'd probably avoid defining CLANG_TRAP_PREFIX as a macro, but instead as an inline or some other form of constant

ahatanak committed:
Define CLANG_TRAP_PREFIX as an inline constexpr StringRef

dwblaikie commented on clang/include/clang/CodeGen/ModuleBuilder.h:
```diff
@@ -28,7 +29,7 @@ namespace llvm {
 
 // Prefix of the name of the artificial inline frame.
-#define CLANG_TRAP_PREFIX "__clang_trap_msg"
+inline constexpr llvm::StringRef CLANG_TRAP_PREFIX = "__clang_trap_msg";
```
The name should be changed to use the usual variable/constant naming convention (ClangTrapPrefix, I think, in this case).

ahatanak committed:
Rename variable

delcypher commented:
@ahatanak Excited to see this finally land 🎉

var-const commented:
@ahatanak Thank you so much for pushing this through! This will be awesome for our hardening effort in libc++. Thanks to everyone for all the feedback and discussions as well!
